From 21b7292db0e9adec9078605613d17bf52067a4bc Mon Sep 17 00:00:00 2001
From: Sudharsan Eswaran <eswaran@qti.qualcomm.com>
Date: Wed, 29 Apr 2026 12:09:41 +0530
Subject: [PATCH] Channel selection 6Ghz Radio issue fix

1. Center Channel to Beacon Channel Conversion Logic fix
2. For 6GHz band get the beacon channel for 40Mhz BW as well
3. modified the hostapd Channel Switch Command for 6GHz
4. Allowed Channel switch even current channel had equal or better preference

Signed-off-by: Sudharsan Eswaran <eswaran@qti.qualcomm.com>
---
 .../slave/tasks/channel_selection_task.cpp    | 73 +++++++++++++------
 .../bwl/nl80211/ap_wlan_hal_nl80211.cpp       | 31 +++-----
 2 files changed, 61 insertions(+), 43 deletions(-)

diff --git a/agent/src/beerocks/slave/tasks/channel_selection_task.cpp b/agent/src/beerocks/slave/tasks/channel_selection_task.cpp
index 6886d8bcd..80b1786e9 100644
--- a/agent/src/beerocks/slave/tasks/channel_selection_task.cpp
+++ b/agent/src/beerocks/slave/tasks/channel_selection_task.cpp
@@ -1612,41 +1612,65 @@ bool ChannelSelectionTask::check_is_there_better_channel_than_current(const sMac
     }
 
     if (current_preference > selected_channel.preference_score) {
+#if 0
         LOG(DEBUG) << "Currect channel is better than the next best channel, no need "
                       "to switch";
         return true;
+#else
+        LOG(DEBUG) << "Currect channel is better - still forcing channel set";
+#endif
     }
 
     // if channel selection request does not explicitly point to a beaconing channel,
     // pick one arbitrarily
     if (son::wireless_utils::is_operating_class_using_central_channel(
-            selected_channel.operating_class) &&
-        son::wireless_utils::is_channel_in_operating_class(selected_channel.operating_class,
-                                                           selected_channel.channel)) {
-        LOG(INFO) << "channel " << int(selected_channel.channel) << " in operating class "
-                  << int(selected_channel.operating_class)
-                  << " needs to be converted to beaconing channel";
+            selected_channel.operating_class)) {
 
         auto freq_type = son::wireless_utils::which_freq_op_cls(selected_channel.operating_class);
 
-        auto beaconing_channels = son::wireless_utils::center_channel_to_beacon_channels(
-            selected_channel.channel, selected_channel.bw, freq_type);
-
-        // if current beaconing channel is in the new chanspec, do not change it;
-        if (std::find(beaconing_channels.begin(), beaconing_channels.end(), current_channel) !=
-            beaconing_channels.end()) {
-
-            selected_channel.channel = current_channel;
-            LOG(INFO) << "new chanspec contains current beaconing channel " << int(current_channel);
-        } else if (beaconing_channels.size() > 1) {
-
-            // in case of 6GHz, this is the Preferred Scanning Channel (PSC)
-            selected_channel.channel = beaconing_channels[1];
-            LOG(INFO) << "new chanspec does not contain current beaconing channel, switch to "
-                      << int(selected_channel.channel);
+        // Check if the selected channel is a center channel or a beacon channel
+        auto center_channel = son::wireless_utils::get_center_channel(
+            selected_channel.channel, freq_type, selected_channel.bw);
+
+        // If get_center_channel returns 0, it means selected_channel.channel is not a valid beacon channel
+        // If it returns a value different from selected_channel.channel, then selected_channel.channel is a beacon channel
+        // If it returns the same value, then selected_channel.channel is a center channel
+        bool is_center_channel = (center_channel != 0 && center_channel == selected_channel.channel);
+
+        if (is_center_channel &&
+            son::wireless_utils::is_channel_in_operating_class(selected_channel.operating_class,
+                                                               selected_channel.channel)) {
+            LOG(INFO) << "channel " << int(selected_channel.channel) << " in operating class "
+                      << int(selected_channel.operating_class)
+                      << " is a center channel, needs to be converted to beaconing channel";
+
+            auto beaconing_channels = son::wireless_utils::center_channel_to_beacon_channels(
+                selected_channel.channel, selected_channel.bw, freq_type);
+
+            // if current beaconing channel is in the new chanspec, do not change it;
+            if (std::find(beaconing_channels.begin(), beaconing_channels.end(), current_channel) !=
+                beaconing_channels.end()) {
+
+                selected_channel.channel = current_channel;
+                LOG(INFO) << "new chanspec contains current beaconing channel " << int(current_channel);
+            } else if (beaconing_channels.size() > 1) {
+
+                // in case of 6GHz, this is the Preferred Scanning Channel (PSC)
+                selected_channel.channel = beaconing_channels[1];
+                LOG(INFO) << "new chanspec does not contain current beaconing channel, switch to "
+                          << int(selected_channel.channel);
+            } else if (beaconing_channels.size() == 1) {
+                selected_channel.channel = beaconing_channels[0];
+                LOG(INFO) << "using the only available beacon channel " << int(selected_channel.channel);
+            } else {
+                LOG(ERROR) << "cannot extract beacon channels for center channel " << int(selected_channel.channel)
+                           << "-" << int(selected_channel.operating_class);
+                return false;
+            }
         } else {
-            LOG(WARNING) << "cannot extract beacon channels for " << int(selected_channel.channel)
-                         << "-" << int(selected_channel.operating_class);
+            LOG(INFO) << "channel " << int(selected_channel.channel) << " in operating class "
+                      << int(selected_channel.operating_class)
+                      << " is already a beacon/primary channel, no conversion needed";
         }
     }
 
@@ -1835,7 +1859,8 @@ ChannelSelectionTask::sSelectedChannel ChannelSelectionTask::select_next_channel
              */
             //
             auto primary_preference = cumulative_preference;
-            if (bandwidth >= eWiFiBandwidth::BANDWIDTH_80) {
+            if (bandwidth >= eWiFiBandwidth::BANDWIDTH_80 ||
+                    (freq_type == beerocks::eFreqType::FREQ_6G && bandwidth >= eWiFiBandwidth::BANDWIDTH_40)) {
                 LOG(INFO) << "[" << channel << "-" << operating_class << "("
                           << beerocks::utils::convert_bandwidth_to_string(bandwidth)
                           << "MHz)] uses a beacon channel.";
diff --git a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
index b8cbe1a49..e724301f1 100644
--- a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
+++ b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
@@ -2174,26 +2174,19 @@ bool ap_wlan_hal_nl80211::switch_channel(int chan, beerocks::eWiFiBandwidth bw,
     }
 
     /*
-    according to the P802.11ax_D7.0 standard, Section 9.4.2.249:
-    On 6GHz band, center_frequency_1 shall be the center frequency of the primary 80MHz channel,
-    and center_frequency_2 shall be the center frequency of the 160MHz channel
+    For 6GHz band with 160MHz:
+    - center_freq1 should be the center frequency of the 160MHz channel
+    - center_freq2 should NOT be used for 160MHz bandwidth
+
+    The P802.11ax_D7.0 standard Section 9.4.2.249 specifies center_freq1 for primary 80MHz
+    and center_freq2 for 160MHz in EHT Operation element, but for the CHAN_SWITCH command
+    to hostapd, we use center_freq1 for the full 160MHz bandwidth.
     */
-    if (freq_type != beerocks::FREQ_6G) {
-        if (bw == beerocks::BANDWIDTH_20) {
-            // For 20MHz, center_freq1 should equal the primary frequency
-            cmd += " center_freq1=" + freq_str;
-        } else if (bw == beerocks::BANDWIDTH_40 || bw == beerocks::BANDWIDTH_80 || bw == beerocks::BANDWIDTH_160) {
-            cmd += " center_freq1=" + wave_vht_center_frequency;
-        }
-    } else {
-        if (bw == beerocks::BANDWIDTH_160) {
-            auto primary_80mhz_freq = (freq < vht_center_frequency) ? vht_center_frequency - 40
-                                                                    : vht_center_frequency + 40;
-            cmd += " center_freq1=" + std::to_string(primary_80mhz_freq);
-            cmd += " center_freq2=" + wave_vht_center_frequency;
-        } else {
-            cmd += " center_freq1=" + wave_vht_center_frequency;
-        }
+    if (bw == beerocks::BANDWIDTH_20) {
+        // For 20MHz, center_freq1 should equal the primary frequency
+        cmd += " center_freq1=" + freq_str;
+    } else if (bw == beerocks::BANDWIDTH_40 || bw == beerocks::BANDWIDTH_80 || bw == beerocks::BANDWIDTH_160) {
+        cmd += " center_freq1=" + wave_vht_center_frequency;
     }
 
     std::string bw_str = beerocks::utils::convert_bandwidth_to_string(bw);
-- 
2.34.1

